### PR TITLE
Add support for rule's handle, position and id

### PR DESCRIPTION
**Pull request by sbezverk**
Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

Having rule's position, handle and ID will help a lot in managing rules.

@stapelberg please let me know what you think about it.

### CODE REVIEW

**Comment by stapelberg** [comment]
Looks great, thanks!

Can you add tests, too?

**Comment by sbezverk** [comment]
@stapelberg Thanks a lot for reviewing. There is a tine architectural issue, currently AddRule actually does not program it and just accumulate it until Flush() programms all the messages. This approach prevents from getting back position and handle which kernel would allocate for a specific rule at the time creation. I am doing some PoC to see how it can be solved with least interference, but most likely AddRule will need to behave the same way as GetRule.

**Comment by sbezverk** [comment]
@stapelberg Please check the latest https://github.com/google/nftables/pull/33/commits/de1205d8691fe0cd1d3537f224001f85e3b1cf3b
commit and for now disregard failed tests.
So far this is the only way I could come up to get rule's handle and position at the creation time. Since it is a breaking change, appreciate your thought about it.

**Comment by sbezverk** [comment]
@stapelberg Please check an alternative version, it does not break existing tests. With this version, `after` rules are programmed into the kernel, `GetRuleHandle` will return actual rule's handle which can be used for further rule management. Please let me know which approach you would prefer.

File: nftables_test.go [hunk 1]
```diff
@@ -624,6 +624,154 @@ func TestTProxy(t *testing.T) {
...
+	if err := c.Flush(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestAddRuleRuleID(t *testing.T) {
+	want := [][]byte{
+		// batch begin
+		[]byte("\x00\x00\x00\x0a"),
+		// nft add rule ip ipv4table ipv4chain-1 position 2 ip version 6
```
Comment (stapelberg) [inline nftables_test.go]: The comment is the same as above, but the byte literal is different. Did you forget to update the comment?

**Comment by stapelberg** [comment]
The non-breaking change is much preferred of course :)

File: nftables_test.go (hunk 1 above)
Comment (sbezverk) [reply nftables_test.go]: oops, sorry you are right, damn copy/paste :)

**Comment by sbezverk** [comment]
@stapelberg fixed the comment, PTAL

File: rule.go [hunk 2]
```diff
@@ -77,32 +81,91 @@ func (cc *Conn) GetRule(t *Table, c *Chain) ([]*Rule, error) {
...
+	rr, err := ruleFromMsg(reply[0])
+	if err != nil {
+		return 0, err
+	}
+
+	return rr.Handle, nil
+}
+
+// AddRule adds the specified Rule
+func (cc *Conn) AddRule(r *Rule) (*Rule, error) {
```
Comment (stapelberg) [inline rule.go]: Can you revert the change to the function signature please? You’re only ever returning nil AFAICT?

Comment (sbezverk) [reply rule.go]: Done

**Comment by sbezverk** [comment]
@stapelberg if you are ok with the change, please approve so I could move forward :)

File: rule.go [hunk 3]
```diff
@@ -77,8 +81,51 @@ func (cc *Conn) GetRule(t *Table, c *Chain) ([]*Rule, error) {
...
-// https://wiki.nftables.org/wiki-nftables/index.php/Simple_rule_management
+// GetRuleHandle returns a specific rule's handle. Rule is identified by Table, Chain and RuleID.
+func (cc *Conn) GetRuleHandle(t *Table, c *Chain, ruleID uint32) (uint64, error) {
+	conn, err := cc.dialNetlink()
+	if err != nil {
+		return 0, err
+	}
+	defer conn.Close()
+	if ruleID == 0 {
+		return 0, fmt.Errorf("Rule's RuleID cannot be 0")
```
Comment (stapelberg) [inline rule.go]: here and below: don’t capitalize error strings: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Comment (sbezverk) [reply rule.go]: done

**Comment by sbezverk** [comment]
@stapelberg I would be very interested in your opinion, please take a look at this one: https://github.com/sbezverk/nftableslib

**Comment by sbezverk** [comment]
@stapelberg any chance you could merge? If everything else is ok of course. Thanks a lot..

**Comment by stapelberg** [comment]
> @stapelberg I would be very interested in your opinion, please take a look at this one: https://github.com/sbezverk/nftableslib

Haven’t looked at it in detail, but on a cursory glance, this looks good!

**Pull request merged**